src/converter/jfr2pprof.java

```diff
@@ -0,0 +1,257 @@
+
+        public Method(final byte[] name) {
+            this.name = name;
+        }
+
+        @Override
+        public int hashCode() {
+            return Arrays.hashCode(name);
+        }
+
+        @Override
+        public boolean equals(final Object other) {
+            if (null == other) {
+                return false;
+            }
+
+            if (other instanceof Method) {
+                final Method that = (Method) other;
+                return Arrays.equals(this.name, that.name);
+            } else {
+                return false;
+            }
+        }
+    }
+
+    private final JfrReader reader;
+
+    public jfr2pprof(final JfrReader reader) {
+        this.reader = reader;
+    }
+
+    public static final byte[] METHOD_UNKNOWN = "[unknown]".getBytes();
+
+    // Profile IDs
+    public static final int PROFILE_SAMPLE_TYPE = 1;
+    public static final int PROFILE_SAMPLE = 2;
+    public static final int PROFILE_LOCATION = 4;
+    public static final int PROFILE_FUNCTION = 5;
+    public static final int PROFILE_STRING_TABLE = 6;
+    public static final int PROFILE_TIME_NANOS = 9;
+    public static final int PROFILE_DURATION_NANOS = 10;
+    public static final int PROFILE_COMMENT = 13;
+    public static final int PROFILE_DEFAULT_SAMPLE_TYPE = 14;
+
+    // ValueType IDs
+    public static final int VALUETYPE_TYPE = 1;
+    public static final int VALUETYPE_UNIT = 2;
+
+    // Sample IDs
+    public static final int SAMPLE_LOCATION_ID = 1;
+    public static final int SAMPLE_VALUE = 2;
+
+    // Location IDs
+    public static final int LOCATION_ID = 1;
+    public static final int LOCATION_LINE = 4;
+
+    // Line IDs
+    public static final int LINE_FUNCTION_ID = 1;
+
+    // Function IDs
+    public static final int FUNCTION_ID = 1;
+    public static final int FUNCTION_NAME = 2;
+
+    // De-dupe string table
+    private int addToStringTable(final Map<byte[], Integer> stringTable, final byte[] value) {
+        final Integer idx = stringTable.get(value);
+        if (null == idx || idx < 0) {
+            final int nextIdx = stringTable.size();
+            stringTable.put(value, nextIdx);
+            return nextIdx;
+        } else {
+            return idx;
+        }
+    }
+
+    // `Proto` instances are mutable, careful with reordering
+    public void dump(final FileOutputStream out) throws Exception {
+        // Mutable IDs, need to start at 1
+        int functionId = 1;
+        int locationId = 1;
+
+        // Used to de-dupe
+        final Map<Method, Long> functions = new HashMap<>();
+        final Map<Method, Long> locations = new HashMap<>();
+
+        final List<ExecutionSample> samples = reader.readAllEvents(ExecutionSample.class);
+        final Dictionary<StackTrace> stackTraces = reader.stackTraces;
+        final long startTicks = reader.startTicks;
+
+        final Map<byte[], Integer> stringTable = new LinkedHashMap<>(); // Will retain order of insertion
```

Comment:
Strictly speaking, the order of insertion does not matter as long as we can later order keys by the corresponding values.
I'm not sure how many strings this map can save at all. If an experiment shows that savings are negligible, I would get rid of this map entirely.